src/Lean/Elab/Tactic/Induction.lean

```diff
@@ -208,15 +208,28 @@ private def getAltNumFields (elimInfo : ElimInfo) (altName : Name) : TermElabM N
 private def isWildcard (altStx : Syntax) : Bool :=
   getAltName altStx == `_
 
-private def checkAltNames (alts : Array Alt) (altsSyntax : Array Syntax) : TacticM Unit :=
+private def checkAltNames (alts : Array Alt) (altsSyntax : Array Syntax) : TacticM Unit := do
+  let mut seenNames : Array Name := #[]
   for h : i in [:altsSyntax.size] do
     let altStx := altsSyntax[i]
     if getAltName altStx == `_ && i != altsSyntax.size - 1 then
       withRef altStx <| throwError "invalid occurrence of wildcard alternative, it must be the last alternative"
     let altName := getAltName altStx
     if altName != `_ then
+      if seenNames.contains altName then
+        throwErrorAt altStx s!"duplicate alternative name '{altName}'"
+      seenNames := seenNames.push altName
       unless alts.any (·.name == altName) do
-        throwErrorAt altStx "invalid alternative name '{altName}'"
+        let unhandledAlts := alts.filter fun alt => !seenNames.contains alt.name
+        let msg ← do
+          if unhandledAlts.isEmpty then
+            pure s!"invalid alternative name '{altName}', no unhandled alternatives"
+          else
+            let unhandledAltsMessages := ((unhandledAlts.map (·.name)).map MessageData.ofName).toList
+            let unhandledAltsStr ← MessageData.toString (MessageData.orList unhandledAltsMessages)
+            pure s!"invalid alternative name '{altName}', expected {unhandledAltsStr}"
+        throwErrorAt altStx msg
```

Comment:
Probably doesn’t matter much, but no need to go through `String` here, is there? Then this becomes non-monadic. Also, repeated `map` can be a single map.

Untested:

```suggestion
        let msg :=
          if unhandledAlts.isEmpty then
            m!"invalid alternative name '{altName}', no unhandled alternatives"
          else
            let unhandledAltsMessages := unhandledAlts.map (m!"{·.name}")
            let unhandledAlts := MessageData.orList unhandledAltsMessages.toList
            m!"invalid alternative name '{altName}', expected {unhandledAlts}"
        throwErrorAt altStx msg
```